### ISSUE TITLE
Panic handler, that saves unsaved work to temporary files

### ISSUE THREAD

**Issue by cessen** [helix-editor/helix#290]
It would be good to have a panic handler that saves the user's work before crashing, so that when a panicking bug is encountered the user doesn't lose all their work.

**Comment by kirawi** [comment]
We could possibly dump the `Document`'s `History` directly to the temporary file since it's unlikely that users will make radically large changes that would negate the space benefit of simply dumping the `History`.

**Comment by archseer** [comment]
I think persistent undo would partly solve this as well

**Comment by cessen** [comment]
@kirawi:
I don't know if optimizing for space/speed/etc. is the right thing to do in this case.  This is an exceptional case (a crash), and I think reliability is most important.

So I would still advocate for simply dumping all unsaved buffers to disk (probably in utf8, regardless of what the original file was).  Basically, just get the user's unsaved work to the disk in the dumbest, hardest-to-screw-up way possible.  And in a format they can easily open with anything, so that if something odd is going on (Helix did crash, after all) they aren't beholden to Helix for conveniently accessing their data again.

In other words, I think this is a feature that should do absolutely *nothing* fancy.  We can save the fancy stuff for when Helix is functioning as expected (which is ideally 100% of the time).

**Comment by cessen** [comment]
> I think persistent undo would partly solve this as well

Yeah, that's true to an extent.  But unless we're really careful about how we implement those kinds of features, things can still potentially get into a weird state (e.g. maybe the on-disk undo history and on-disk file are out of sync, and it's hard to figure out where to start in the undo history to correctly recover the lost data).

This feature is basically like wearing your seat belt: you don't expect to crash, but just in case you do, you'd like to minimize irrecoverable damage.  And I think it should be pretty simple to implement, at least to the extent that we're willing to trust Ropey to always stay in a valid state.

**Comment by pickfire** [comment]
Maybe swap file will do? For every change we keep a swap file in /tmp then if it panics we have a panic handler to report the swap file within /tmp.

**Comment by kirawi** [comment]
@cessen How about file compression then? If I was editing a big file for whatever reason and Helix crashed, I wouldn't want it to write it completely down.

**Comment by cessen** [comment]
> Maybe swap file will do?

Yeah, we could do that.  Basically like auto-save "light".  Then on crash, the swap file will always still be there.  That's probably more robust than the approach I've been suggesting.

> How about file compression then?

The key point I'm trying to make is that this code will only run in exceptional circumstances, and ideally it will never run at all.  It's also code that only runs when something has gone wrong (potentially seriously wrong).

All of these factors mean that it should be the dumbest, hardest-to-screw-up thing possible.  Every additional layer of *anything* that we add on top of it is another code path that may have its own bugs.  And for something that is supposed to be the safeguard in case of crashes, minimizing the potential for bugs trumps even major efficiency gains, I think.

Usually I'm all for making things faster, more efficient, etc.  But this is one of the few cases where I think that's basically a non-consideration.

**Comment by cessen** [comment]
Over the last day or so I used cargo-fuzz to fuzz the mutation methods on Ropey's `Rope` type.  I let it run for ~14 hours on my 24-core workstation, for a total of ~1 billion iterations.

Each iteration also includes a check that panics if the `Rope` doesn't pass any of Ropey's own validation checks.  So the fuzzing is effectively searching not just for unexpected panics/memory-safety-violations/etc., but also for anything that can put the `Rope` into an invalid state by Ropey's own definition.

It found a single issue within the first 10 minutes or so (a benign validation-check violation), which I've fixed.  But other than that, it's found zero issues.  That's not proof of perfection, of course.  But it does indicate that Ropey's `Rope` type is at least very difficult to get into an invalid state.  So I think we can reasonably depend on it, if we want to go the route of dumping from memory.

**Comment by kirawi** [comment]
Should we prefer #401 to this?

**Comment by antoyo** [comment]
> Should we prefer #401 to this?

I personally see those as two different features: for instance, I'd probably want to disable persistent undo as this is not a feature I want yet I'd like to have swap files in case of problems.

**Comment by nilueps** [comment]
#4215 author here. How difficult would you estimate this fix would be to implement? I'm a more or less advanced beginner when it comes to rust and i'd be interested in tackling this issue if it isn't too far out of my league.

**Comment by Philipp-M** [comment]
> #4215 author here. How difficult would you estimate this fix would be to implement? I'm a more or less advanced beginner when it comes to rust and i'd be interested in tackling this issue if it isn't too far out of my league.

I think first a decision about how this should be implemented should be made. I also lean towards a general solution as #401.
I don't think that this necessarily means persistent undo, but rather some kind of general session state that may be configurable, that is serialized to disk also in a panic handler (at least, if possible without serializing inconsistent state).

**Comment by josephholsten** [comment]
I think we're missing the fact that most process crashes aren't due to internal panic but external factors, ie unexpected system shutdown, out-of-memory, some idiot forgetting his laptop is set to automatically log out after an hour of inactivity.

So while a panic handler is great, a swap file or operation journal is vital feature.

**Comment by 6lj6** [comment]
So there is no method now to get my lost file back?

**Comment by WarpspeedSCP** [comment]
I'd really appreciate a swap file feature; I just lost about an hour's work because I assumed helix would at least hold onto my unsaved work when I closed a console tab (I mean, even vim does it!)

**Comment by vgwidt** [comment]
I'm too erratic to operate without swap files and was bitten pretty quick after I began to use Helix, so I tried to implement a basic swap file system to see what it could look like (https://github.com/vgwidt/helix/tree/master).

A swap file (.<filename>.hxs) is created in the same folder as the file (path).  Saves are written to the swap file then the file is copied to the original path.  The swap is removed on close.  If a swap file exists when opening a file, it tells the user to fix it (no options yet for choosing how to handle).  It does not create a swap for scratch files.  I attempted auto save using the aggressive idle-timeout (for testing only) but didn't have success, so right now the only advantage it really provides is protection on crash during a save (#3967).  It could easily fix the issues described in #4257 as well.

There would be some performance implications here if implemented this way.  I think Vim swap files record changes which are merged on write, along with some other metadata.  It might be nice to do something similar.  I think Vim also saves the swap every 4 idle seconds or after certain amount of inserted characters.

I don't know if this is the route that will be taken, but I'll try to get the auto save working for the swap file and maybe try to clean it up a bit.

**Comment by bcspragu** [comment]
Similar to @sqwxl, I'm happy to contribute here and care mostly about the "something went wrong (internally or externally) and the program went down hard, where did the last `<duration>` of my work end up?". Especially when building from `main`, I've hit ([and](https://github.com/helix-editor/helix/issues/3459) [filed](https://github.com/helix-editor/helix/issues/6752)) a number of hard crashes that cost me a lot of time. On the bright side, I'm now a compulsive `:w`-er.

I'm not opinionated on the exact fix (swap files, dumping changesets, etc), just opinionated that there should be a safety net in place to recover work after a crash.

**Comment by kirawi** [comment]
Part of the problem is that there is a bug that prevents swap files from working: https://github.com/helix-editor/helix/pull/4276. It's hard to create a minimal reproducible case as well.

**Comment by pascalkuthe** [comment]
I also don't really see how this would actually work.

A panic handler only has access to statics so we would need to store our document data in a static which would require locking on every transaction (which I don't see as acceptable).

I think one way to handle this ks to run the main UI loop in it's own thread outside of the tokio executer (tokio recommends that anyway) and have editor as a thread local (it's non-sync anyway). That way the panic handler could access it. But that's a pretty large refactor

**Comment by kirawi** [comment]
I think that https://github.com/helix-editor/helix/pull/4276 (if the bug is found) + persistent undo would cover this.

**Comment by pascalkuthe** [comment]
persistant undo would work but you wouldn't want to write persistent undo to disk on every keypress either. So you would still need to write to diisk at some point. That point would be when saving so that doens't dicetly solve it until we save to disk in the panic handler (or write the undofile).

That shill has the problem that we don't have access to either undofile or document state in the panic handler (since that can only access statics.

You might work around this with catch unwind instead I suppose but I am not sure how the tokio runtime handles panics (I think the entire runtime just shuts down so you can't catch panics in a a different worker?)

**Comment by kirawi** [comment]
How does vim/kakoune handle this, if at all?

**Comment by pascalkuthe** [comment]
Vim only has lockfiles that are auto saved after 4 seoncds of no changes. Since C doesn't really have exceptions writing a proper crash handler there is hard. Once we implement autosave we could simply offer three options:no autafe, swap file and enable.

Kakoune doesn't have swapfiles but since c++ does have exceptions (altough it still can and does hard crash too) they catch exceptions and write all buffers to disk. That is essentially the same I said above but it's harder to do in rust/with how we handle things. 

It seems catching the panic may actually work tough.

**Comment by ardijanr** [comment]
New to helix and just got bitten by this myself, lost about 2 hours of work. Idk why I just assumed I would not loose the work. 

If it can be done with panic then I guess I am all for it, but I somehow doubt you could make it as robust as "save every X seconds". Good luck saving when I am holding down the power button, or my server looses mains. 

My opinion is that helix should still be able to recover at least some of the work.

I would suggest that swap files and history state is saved in the same location. 

If this is still open when I have time, I will give it a go.

**Comment by kirawi** [comment]
See the above linked PRs that already do essentially the same thing.

**Comment by ardijanr** [comment]
#4276 seems to deal with fixing manual saves. 
I would personally prefer #4257, which is why I am writing here since it seems allot of people seem to think that this would replace swap files. Which imo it would not.

The only other PR/issue which I find relevant above is issue #401. And if I understand you correctly, you want #401 and just panic saves?

I guess panic and history recovery could work, idk how this has been implemented so I am curious as to how you handle:

- Extremely long history.
- File is never saved manually.
- History fails to apply or load.

Relying only on history to recover a file seems more prone to failure than just keeping a backup of the file which is deleted once the changes are written to the original file.

IMO recoverable history is separate and does only that; recover change history such that undo works. It can fail and if it does it should not be a "really" big deal. In other words I did not loose a significant amount of changes I just lost history. But maybe you have the entire buffer in the history file too, such that you can at least recover that bit. Maybe recoverable history works more like a backup, I honestly don't know since I cannot seem to find which branch it would be at.

Also maybe I am just blind as to how panic save works but I cannot see how panic save can write a file when power to the system is lost unexpectedly.
